Refactoring TextDecorationLine to fit in 5 bits

vitorroriz opened the pull request:
#### 02d04382a054cfe1edcb38c58ba1c1d7795eeefb
<pre>
Refactoring TextDecorationLine to fit in 5 bits
<a href="https://bugs.webkit.org/show_bug.cgi?id=297940">https://bugs.webkit.org/show_bug.cgi?id=297940</a>
<a href="https://rdar.apple.com/159236131">rdar://159236131</a>

Reviewed by Sammy Gill.

At <a href="https://commits.webkit.org/299142@main">https://commits.webkit.org/299142@main</a> we have refactored TextDecorationLine
into a strong type. However, it increased its size to 2 bytes and it had to be
moved out from RenderStyle Flags.

We have now refactored it, packing its content within 5 bits, so we can move
it back to RenderStyle flags where it used to be.

We are representing TextDecorationLine in 5 bits.
1 bit is used for defining the Type (SingleValue or Flags)
4 bits are used for defining the Value
Values for SingleValue: None, SpellingError, GrammarError
Values for Flags: Any combination of Underline, Overline, LineThrough, Blink
Therefore, we are packing its content with the following layout:
Bits 7-5: Reserved
Bit 4: Type (SingleValue or Flags)
Bits 3-0: When Type=1 (Underline=0x1, Overline=0x2, LineThrough=0x4, Blink=0x8)
         : When Type=0 (None = 0, SpellingError = 1, GrammarError = 2)

* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::RenderStyle):
(WebCore::RenderStyle::NonInheritedFlags::copyNonInheritedFrom):
(WebCore::RenderStyle::changeAffectsVisualOverflow const):
(WebCore::miscDataChangeRequiresRepaint):
(WebCore::RenderStyle::changeRequiresRepaint const):
(WebCore::RenderStyle::changeRequiresRepaintIfText const):
(WebCore::RenderStyle::conservativelyCollectChangedAnimatableProperties const):
(WebCore::RenderStyle::NonInheritedFlags::dumpDifferences const):
(WebCore::RenderStyle::InheritedFlags::dumpDifferences const):
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
(WebCore::RenderStyle::textDecorationLine const):
(WebCore::RenderStyle::textDecorationLineInEffect const):
* Source/WebCore/rendering/style/RenderStyleSetters.h:
(WebCore::RenderStyle::addToTextDecorationLineInEffect):
(WebCore::RenderStyle::setTextDecorationLine):
(WebCore::RenderStyle::setTextDecorationLineInEffect):
* Source/WebCore/rendering/style/StyleInheritedData.cpp:
(WebCore::StyleInheritedData::StyleInheritedData):
(WebCore::StyleInheritedData::nonFastPathInheritedEqual const):
(WebCore::StyleInheritedData::dumpDifferences const):
* Source/WebCore/rendering/style/StyleInheritedData.h:
* Source/WebCore/rendering/style/StyleMiscNonInheritedData.cpp:
(WebCore::StyleMiscNonInheritedData::StyleMiscNonInheritedData):
(WebCore::StyleMiscNonInheritedData::operator== const):
(WebCore::StyleMiscNonInheritedData::dumpDifferences const):
* Source/WebCore/rendering/style/StyleMiscNonInheritedData.h:
* Source/WebCore/style/values/text-decoration/StyleTextDecorationLine.cpp:
(WebCore::Style::TextDecorationLine::addOrReplaceIfNotNone):
(WebCore::Style::operator&lt;&lt;):
* Source/WebCore/style/values/text-decoration/StyleTextDecorationLine.h:
(WebCore::Style::TextDecorationLine::TextDecorationLine):
(WebCore::Style::TextDecorationLine::type const):
(WebCore::Style::TextDecorationLine::isNone const):
(WebCore::Style::TextDecorationLine::isSpellingError const):
(WebCore::Style::TextDecorationLine::isGrammarError const):
(WebCore::Style::TextDecorationLine::isFlags const):
(WebCore::Style::TextDecorationLine::hasUnderline const):
(WebCore::Style::TextDecorationLine::hasOverline const):
(WebCore::Style::TextDecorationLine::hasLineThrough const):
(WebCore::Style::TextDecorationLine::hasBlink const):
(WebCore::Style::TextDecorationLine::containsAny const):
(WebCore::Style::TextDecorationLine::contains const):
(WebCore::Style::TextDecorationLine::remove):
(WebCore::Style::TextDecorationLine::switchOn const):
(WebCore::Style::TextDecorationLine::toRaw const):
(WebCore::Style::TextDecorationLine::operator== const):
(WebCore::Style::TextDecorationLine::rawValue const):
(WebCore::Style::TextDecorationLine::packFlags):
(WebCore::Style::TextDecorationLine::packFlag):
(WebCore::Style::TextDecorationLine::unpackFlags const):

Canonical link: <a href="https://commits.webkit.org/299294@main">https://commits.webkit.org/299294@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/052d8c9271bf975041d93101802119913746c4f2

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/118533 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/38214 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/28865 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/124705 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/70592 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/9f21edfb-c6cb-4758-b59b-dd4492f50fe3) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/120411 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/38910 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/46796 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/89968 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/70592 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/bc052dd5-9a9f-4b45-bd41-5841789b756c) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/121486 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/30995 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/106278 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/70472 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/552ee652-39b6-4dfc-8725-31004137a20e) 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/30052 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/24389 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/68364 "Built successfully") | | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/100434 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/24580 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/127771 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/45440 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/34280 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/127771 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/45804 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/102499 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/127771 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25009 "Built successfully and passed tests") | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/126/builds/43823 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/21818 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/41938 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/45310 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/50988 "Built successfully") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/44773 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/48120 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/46460 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

sammygill commented on Source/WebCore/style/values/text-decoration/StyleTextDecorationLine.h:
```diff
@@ -28,116 +28,224 @@
+    static constexpr uint8_t BlinkBit = static_cast<uint8_t>(TextDecorationLineFlags::Blink);
+
+    // values when Type is SingleValue
+    enum class SingleValues :uint8_t {
```
Probably want a space after the `:`

sammygill replied:
Also should this be `SingleValue` instead of `SingleValues`?

sammygill commented on Source/WebCore/style/values/text-decoration/StyleTextDecorationLine.h:
```diff
@@ -28,116 +28,224 @@
+    static constexpr uint8_t SingleValueNone          = static_cast<uint8_t>(Type::SingleValue) | static_cast<uint8_t>(SingleValues::None);
+    static constexpr uint8_t SingleValueSpellingError = static_cast<uint8_t>(Type::SingleValue) | static_cast<uint8_t>(SingleValues::SpellingError);
+    static constexpr uint8_t SingleValueGrammarError  = static_cast<uint8_t>(Type::SingleValue) | static_cast<uint8_t>(SingleValues::GrammarError);
+
```
Extra newline?

sammygill commented on Source/WebCore/style/values/text-decoration/StyleTextDecorationLine.h:
```diff
@@ -28,116 +28,224 @@
+    // Returns only the value bits, not to be confused with "toRaw", which returns the whole packed raw representation
+    inline uint8_t rawValue() const { return m_packed & ValuesMask; }
+
+    static constexpr uint8_t packFlag(TextDecorationLineFlags flag)
```
I know this is not part of your change but I wonder if this should also be singular. It's a bit confusing that the type is plural but it's only actually a single flag we are packing here

vitorroriz replied:
I can rename it but it has 90 occurrences, should I do it here?

sammygill replied:
Probably not I just mentioned it as something to think about in an additional patch :)

sammygill commented on Source/WebCore/style/values/text-decoration/StyleTextDecorationLine.h:
```diff
@@ -28,116 +28,224 @@
+    static constexpr uint8_t packFlag(TextDecorationLineFlags flag)
+    {
+        switch (flag) {
+        case TextDecorationLineFlags::Underline:
+            return UnderlineBit;
+        case TextDecorationLineFlags::Overline:
+            return OverlineBit;
+        case TextDecorationLineFlags::LineThrough:
+            return LineThroughBit;
+        case TextDecorationLineFlags::Blink:
+            return BlinkBit;
```
Should we be setting the "Flags," bit here too? I see in many of the contains functions we check `isFlags()` and then check the values. If we set that here maybe we could just look at the whole packed expression and check the set bits

sammygill commented on Source/WebCore/style/values/text-decoration/StyleTextDecorationLine.h:
```diff
@@ -28,116 +28,221 @@
+    static constexpr uint8_t TypeMask = 1 << 4; // 0001 0000
+    static constexpr uint8_t ValuesMask = 0x0F;
+
+    // values when Type is SingleValue
```
Nit: First letter should probably be capitalized to match the other comments

sammygill commented on Source/WebCore/style/values/text-decoration/StyleTextDecorationLine.h:
```diff
@@ -28,116 +28,221 @@
+        GrammarError
+    };
+
+    // values when Type is Flags
```
Nit: First letter should probably be capitalized to match the other comments

vitorroriz committed:
Refactoring TextDecorationLine to fit in 5 bits
https://bugs.webkit.org/show_bug.cgi?id=297940
rdar://159236131

Reviewed by Sammy Gill.

At https://commits.webkit.org/299142@main we have refactored TextDecorationLine
into a strong type. However, it increased its size to 2 bytes and it had to be
moved out from RenderStyle Flags.

We have now refactored it, packing its content within 5 bits, so we can move
it back to RenderStyle flags where it used to be.

We are representing TextDecorationLine in 5 bits.
1 bit is used for defining the Type (SingleValue or Flags)
4 bits are used for defining the Value
Values for SingleValue: None, SpellingError, GrammarError
Values for Flags: Any combination of Underline, Overline, LineThrough, Blink
Therefore, we are packing its content with the following layout:
Bits 7-5: Reserved
Bit 4: Type (SingleValue or Flags)
Bits 3-0: When Type=1 (Underline=0x1, Overline=0x2, LineThrough=0x4, Blink=0x8)
         : When Type=0 (None = 0, SpellingError = 1, GrammarError = 2)

* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::RenderStyle):
(WebCore::RenderStyle::NonInheritedFlags::copyNonInheritedFrom):
(WebCore::RenderStyle::changeAffectsVisualOverflow const):
(WebCore::miscDataChangeRequiresRepaint):
(WebCore::RenderStyle::changeRequiresRepaint const):
(WebCore::RenderStyle::changeRequiresRepaintIfText const):
(WebCore::RenderStyle::conservativelyCollectChangedAnimatableProperties const):
(WebCore::RenderStyle::NonInheritedFlags::dumpDifferences const):
(WebCore::RenderStyle::InheritedFlags::dumpDifferences const):
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
(WebCore::RenderStyle::textDecorationLine const):
(WebCore::RenderStyle::textDecorationLineInEffect const):
* Source/WebCore/rendering/style/RenderStyleSetters.h:
(WebCore::RenderStyle::addToTextDecorationLineInEffect):
(WebCore::RenderStyle::setTextDecorationLine):
(WebCore::RenderStyle::setTextDecorationLineInEffect):
* Source/WebCore/rendering/style/StyleInheritedData.cpp:
(WebCore::StyleInheritedData::StyleInheritedData):
(WebCore::StyleInheritedData::nonFastPathInheritedEqual const):
(WebCore::StyleInheritedData::dumpDifferences const):
* Source/WebCore/rendering/style/StyleInheritedData.h:
* Source/WebCore/rendering/style/StyleMiscNonInheritedData.cpp:
(WebCore::StyleMiscNonInheritedData::StyleMiscNonInheritedData):
(WebCore::StyleMiscNonInheritedData::operator== const):
(WebCore::StyleMiscNonInheritedData::dumpDifferences const):
* Source/WebCore/rendering/style/StyleMiscNonInheritedData.h:
* Source/WebCore/style/values/text-decoration/StyleTextDecorationLine.cpp:
(WebCore::Style::TextDecorationLine::addOrReplaceIfNotNone):
(WebCore::Style::operator<<):
* Source/WebCore/style/values/text-decoration/StyleTextDecorationLine.h:
(WebCore::Style::TextDecorationLine::TextDecorationLine):
(WebCore::Style::TextDecorationLine::type const):
(WebCore::Style::TextDecorationLine::isNone const):
(WebCore::Style::TextDecorationLine::isSpellingError const):
(WebCore::Style::TextDecorationLine::isGrammarError const):
(WebCore::Style::TextDecorationLine::isFlags const):
(WebCore::Style::TextDecorationLine::hasUnderline const):
(WebCore::Style::TextDecorationLine::hasOverline const):
(WebCore::Style::TextDecorationLine::hasLineThrough const):
(WebCore::Style::TextDecorationLine::hasBlink const):
(WebCore::Style::TextDecorationLine::containsAny const):
(WebCore::Style::TextDecorationLine::contains const):
(WebCore::Style::TextDecorationLine::remove):
(WebCore::Style::TextDecorationLine::switchOn const):
(WebCore::Style::TextDecorationLine::toRaw const):
(WebCore::Style::TextDecorationLine::operator== const):
(WebCore::Style::TextDecorationLine::rawValue const):
(WebCore::Style::TextDecorationLine::packFlags):
(WebCore::Style::TextDecorationLine::packFlag):
(WebCore::Style::TextDecorationLine::unpackFlags const):

Canonical link: https://commits.webkit.org/299294@main
